lib/galaxy/tools/toolbox/base.py

```diff
@@ -79,7 +79,13 @@ def __init__( self, config_filenames, tool_root_dir, app, tool_conf_watcher=None
         if tool_conf_watcher:
             self._tool_conf_watcher = tool_conf_watcher  # Avoids (re-)starting threads in uwsgi
         else:
-            self._tool_conf_watcher = get_tool_conf_watcher(lambda: self.handle_reload_toolbox())
+            if hasattr(self.app, 'tool_cache'):
+                # Normal galaxy instances should have a tool_cache,
+                # but the toolshed does not.
+                tool_cache = self.app.tool_cache
+            else:
+                tool_cache = None
+            self._tool_conf_watcher = get_tool_conf_watcher(reload_callback=lambda: self.handle_reload_toolbox(), tool_cache=tool_cache)
```

Comment:
Actually after posting the previous comment I thought that maybe it would be better to have a single toolbox watcher monitoring tool and data manager config files plus the tool_cache, i.e. get rid of the data manager separate tool conf watcher. It seems we are on the same line!